lib/llhttp/parser.js

```diff
@@ -0,0 +1,240 @@
+'use strict'
+
+/* global WebAssembly */
+
+const { resolve } = require('path')
+const { readFileSync } = require('fs')
+const constants = require('./constants')
+const { kMaxHeadersSize } = require('../core/symbols')
+const WASM_BUILD = resolve(__dirname, './llhttp.wasm')
+const bin = readFileSync(WASM_BUILD)
+const mod = new WebAssembly.Module(bin)
+
+const kOnMessageBegin = 0
+const kOnHeaders = 1
+const kOnHeadersComplete = 2
+const kOnBody = 3
+const kOnMessageComplete = 4
+const kOnExecute = 5
+
+const kPtr = Symbol('kPrt')
+const kStatusMessage = Symbol('kStatusMessage')
+const kHeadersFields = Symbol('kHeadersFields')
+const kHeadersValues = Symbol('kHeadersValues')
+const kBufferSize = Symbol('kBufferSize')
+const kHeadersTimeout = Symbol('kHeadersTimeout')
+const kBufferPtr = Symbol('kBufferPtr')
+const kMallocBuffer = Symbol('kMallocBuffer')
+const kCurrentBuffer = Symbol('kCurrentBuffer')
+const kGetHeaders = Symbol('kGetHeaders')
+const kMakeError = Symbol('kMakeError')
+
+/**
+ * Map of HTTPParser instances
+ */
+const instMap = new Map()
+
+const cstr = (ptr, len) =>
+  Buffer.from(inst.exports.memory.buffer, ptr, len).toString()
+
+/* eslint-disable camelcase */
+const wasm_on_message_begin = p => {
+  const i = instMap.get(p)
+  this[kStatusMessage] = null
+  return i[kOnMessageBegin]()
+}
+
+const wasm_on_url = (p, at, length) => {
+  return 0
+}
+
+const wasm_on_status = (p, at, length) => {
+  instMap.get(p)[kStatusMessage] = cstr(at, length)
+  return 0
+}
+
+const wasm_on_header_field = (p, at, length) => {
+  instMap.get(p)[kHeadersFields].push(cstr(at, length))
+  return 0
+}
+
+const wasm_on_header_value = (p, at, length) => {
+  instMap.get(p)[kHeadersValues].push(cstr(at, length))
+  return 0
+}
+
+const wasm_on_headers_complete = p => {
+  const i = instMap.get(p)
+  const versionMajor = inst.exports.llhttp_get_http_major(p)
+  const versionMinor = inst.exports.llhttp_get_http_minor(p)
+  const rawHeaders = i[kGetHeaders]()
+  const statusCode = inst.exports.llhttp_get_status_code(p)
+  const statusMessage = i[kStatusMessage]
+  const upgrade = Boolean(inst.exports.llhttp_get_upgrade(p))
+  const shouldKeepAlive = inst.exports.llhttp_should_keep_alive(p)
```

Comment:
What is the type of upgrade and shouldKeepAlive in the node native parser?

Reply:
`shouldKeepAlive` should be boolean (https://github.com/nodejs/node/blob/master/src/node_http_parser.cc#L358)

`upgrade` too (https://github.com/nodejs/node/blob/master/src/node_http_parser.cc#L363)

I think they are `0`/`1` in `llhttp`. I saw `undici`'s parser handled it. I'll check again.

Reply:
Shouldn't you case shouldKeepAlive to a Boolean as well then?

Reply:
I'll check it, maybe I'll have to 👍🏻